README.md
# chalkbars

[![Gem Version](https://badge.fury.io/js/chalkbars.png)](http://badge.fury.io/js/chalkbars)
[![Dependency Status](https://gemnasium.com/ShogunPanda/chalkbars.png?travis)](https://gemnasium.com/ShogunPanda/chalkbars)
[![Build Status](https://secure.travis-ci.org/ShogunPanda/chalkbars.png?branch=master)](http://travis-ci.org/ShogunPanda/chalkbars)
[![Coverage Status](https://coveralls.io/repos/ShogunPanda/chalkbars/badge.png)](https://coveralls.io/r/ShogunPanda/chalkbars)

Template based terminal coloring based on [Chalk](https://www.npmjs.com/package/chalk) and [Handlebars](http://handlebarsjs.com).

http://sw.cowtech.it/chalkbars

## Usage

Chalkbars allows an easier use of chalk on the console taking advantage of the awesome Handlebars templating engine.

The use is really easy:

```javascript
var chalkbars = require("chalkbars").format;
console.log(chalkbars("{{#C red}}This is in red{{/C}} and this is not."));
```

## Supported implementations.

Chalkbars supports and has been tested on [NodeJS](http://nodejs.org) 5.0+. 

## Chalkbars Handlebars helpers

### C

This is the main template. It accepts one or more styles, concatenated by any character except letters, numbers and "#". 

Even thought quotes are optional, is advised to always provide them for clarity.

This template will replace text inside with the styles provided. 

It supports nesting.

Examples:

```javascript
console.log(chalkbars("{{#C red}}This is in red except for {{#C green}}this{{/C}} and it supports restoring.{{/C}}"));
console.log(chalkbars("{{#C red.bgYellow}}This is in red with a yellow background.{{/C}}"));
console.log(chalkbars("{{#C red$bgYellow}}Same as above.{{/C}}"));
```

Supported styles are:

* Default Chalk styles.
* _i#_**ABC** (where ABC is a 3 digit number): The ANSI color code. Example: **031** is red foreground, **042** is green background.
* _x#_**ABC** (where A, B, C are a number from 0 to 5): The 256 ANSI color code. For the meaning of the values, see [ansi-256-colors](https://github.com/jbnicolai/ansi-256-colors).
* _X#_**AABBCC** (where AA, BB, CC ranges from 00 to FF): The RGB HEX color code.
* User defined styles (see ```chalkbars.style``` below).

All unrecognized styles are ignored.

### B

Outputs a header with a word between brackets. 

It does not support nesting and does not support any embedded content.
 
It accepts one or two arguments. The first argument is the word to print, the second argument is the style. If the latter is omitted, the first word is also used as a style.
Again, quoting is optional but advised.

Examples: 

```javascript
console.log(chalkbars("{{B info}} On the left you see a white 'INFO' word inside blue brackets"));
console.log(chalkbars("{{B \"info\" \"red\"}} On the left you see a red 'INFO' word inside blue brackets"));
console.log(chalkbars("{{B info red}} Omitting quotes might lead to unwanted behavior"));
console.log(chalkbars("{{B \"what\" \"red.bgYellow\"}} On the left you see a red on yellow background 'WHAT' word inside blue brackets"));
```

#### Shortcuts

The following shortcut helpers are also defined:

* **BI**: Shortand for `{{B "info"}}`
* **BW**: Shortand for `{{B "warn"}}`
* **BO**: Shortand for `{{B "ok"}}`
* **BF**: Shortand for `{{B "fail"}}`
* **BE**: Shortand for `{{B "error"}}`
* **BP**: Shortand for `{{B "pass"}}`
* **BS**: Shortand for `{{B "skip"}}`
* **BD**: Shortand for `{{B "debug"}}`

### E

Outputs a footer with a word between brackets

It behaves like [B](#user-content-b) but it makes sure the message is printed to the right of the screen.

```javascript
console.log(chalkbars("On the right you see a white 'INFO' word inside blue brackets {{E info}}"));
console.log(chalkbars("On the right you see a red 'INFO' word inside blue brackets {{E \"info\" \"red\"}}"));
console.log(chalkbars("Omitting quotes might lead to unwanted behavior {{E info red}}"));
console.log(chalkbars("On the right you see a red on yellow background 'WHAT' word inside blue brackets {{B \"what\" \"red.bgYellow\"}}"));
```

#### Shortcuts

The following shortcut helpers are also defined:

* **EI**: Shortand for `{{E "info"}}`
* **EW**: Shortand for `{{E "warn"}}`
* **EO**: Shortand for `{{E "ok"}}`
* **EF**: Shortand for `{{E "fail"}}`
* **EE**: Shortand for `{{E "error"}}`
* **EP**: Shortand for `{{E "pass"}}`
* **ES**: Shortand for `{{E "skip"}}`
* **ED**: Shortand for `{{E "debug"}}`

## Configuring Chalkbars

### chalkbars.configuration.silent

Whether to ignore exceptions when compiling a Handlebars template fails. 

If set to `false` a compilation failure will raise an exception.

The default value is `true`.

### chalkbars.configuration.openingBracket

The default template for the opening bracket in the [B](#user-content-b) and [E](#user-content-e) templates.

The default value is `{{#C bracket}}[{{/C}}`.

### chalkbars.configuration.closingBracket

The default template for the closing bracket in the [B](#user-content-b) and [E](#user-content-e) templates.

The default value is `{{#C bracket}}]{{/C}}`.

### chalkbars.configuration.styles

The list of valid chalk styles for the [C](#user-content-c), [B](#user-content-b) and [E](#user-content-e) templates.  

**Do not modify this directly but use [chalkbars.style](#user-content-chalkbarsstylename-value) instead.**

## API Documentation

### chalkbars.format(...template, context)

It compiles the template (you can pass multiple templates, they will be concatenated) using Handlebars and applying chalk and custom styles.

If the last argument is a object, it will be passed as the context to Handlebars.

If might or not raise an exception depending on the value of `chalkbars.configuration.silent`.

Example:

```javascript
console.log(chalkbars.format("{{#C red}}This is in red{{/C}} and this is not."));
```

### chalkbars.style(name, value)

Get or sets a chalkbars style. Returns `undefined` when a style is not defined. 

When `value` is omitted, it returns the current value of the style.

When `value` is `null`, it returns the current value of the style and then deletes the style.
     
When `value` is a string, it sets the new value of the style and then returns the value itself.

Style nesting or overwriting a chalk style is not allowed.
 
Examples: 

```javascript
console.log(chalkbars.style("custom"));
// => undefined
console.log(chalkbars.style("custom", "red"));
// => "red"
console.log(chalkbars.style("custom"));
// => "red"
console.log(chalkbars.style("custom", "green"));
// => "green"
console.log(chalkbars.style("custom", null));
// => "green"
console.log(chalkbars.style("custom"));
// => undefined
```

### chalkbars.log(...template, context)

Simply a shortcut for `console.log(chalkbars.format(...));`.

### chalkbars.formatNoColor(...template, context)

It behaves like [chalkbars.format](#user-content-chalkbarsformattemplate-context) but it removes all ANSI colors escapes code before returning.

### chalkbars.plainFormat(...template, context)

It behaves like [chalkbars.format](#user-content-chalkbarsformattemplate-context) but it removes **all** ANSI escapes code before returning.

### chalkbars.chalk

The console library used by chalkbars. More information [here](https://www.npmjs.com/package/chalk).

### chalkbars.handlebars

The templating library used by chalkbars. More information [here](http://handlebarsjs.com).

## Contributing to chalkbars

* Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet.
* Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it.
* Fork the project.
* Start a feature/bugfix branch.
* Commit and push until you are happy with your contribution.
* Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.

## Copyright

Copyright (C) 2015 and above Shogun (shogun@cowtech.it).

Licensed under the MIT license, which can be found at http://opensource.org/licenses/MIT.


CHANGELOG.md
### 2016-03-23 / 2.0.0

* **Dropped support for Node < 5.**
* Added `chalkbars.error`, which is a shorthand for `console.error(chalkbars.format(...))`.
* Update versions of Chalk and Handlebars.

### 2015-04-05 / 1.1.0

* Export handlebars and chalk to allow template registration.

### 2015-04-05 / 1.0.1 - The Easter release!

* Changed default colors for debug and highlight.

### 2015-03-14 / 1.0.0 - The PI release!

* Added JSDoc documentation.
* Written README.
* First public version.

### 2015-03-14 / 0.3.0 (not public)

* Renamed `$` to `format`.
* Added `chalkbars.formatNoColor`, which compiles template removing all ANSI colors.
* Added `chalkbars.plainFormat`, which compiles template removing all ANSI escapes.
* Added `chalkbars.log`, which is a shorthand for `console.log(chalkbars.format(...))`.
* Improved tests.
* Added ESLint for QA.

### 2015-02-22 / 0.2.2 (not public)

* Bugfix.
* Added Travis configuration.

### 2015-02-22 / 0.2.1 (not public)

* Bugfix.

### 2015-02-22 / 0.2.0 (not public)

* Renamed `cb` helper to `C`.
* Renamed `cbh` helper to `B`.
* Renamed `cbf` helper to `E`.
* Added shortcuts for common usages of `B` and `E` styles (see `configuration.js`).

### 2015-01-17 / 0.1.0 (not public)

* Initial version.


test/test.js
/*
 * This file is part of the chalkbars npm package. Copyright (C) 2015 and above Shogun <shogun@cowtech.it>.
 * Licensed under the MIT license, which can be found at http://www.opensource.org/licenses/mit-license.php.
 */

/* globals describe, it */
/* eslint-disable no-unused-expressions */

const expect = require("chai").expect;
const sinon = require("sinon");
const chalk = require("chalk");
const chalkbars = require("../main");

const verifyFormat = function(){
  const args = Array.prototype.slice.call(arguments);
  const expected = args.pop();

  expect(chalkbars.format.apply(this, args)).to.equal(expected);
};

process.stdout.columns = 80;

describe("chalkbars.format", () => {
  describe("arguments handling", () => {
    it("should return a empty string when no arguments are given", () => {
      verifyFormat("");
      verifyFormat("", "");
    });

    it("should style using arguments and invoking chalk", () => {
      verifyFormat("{{#C \"red\"}}MESSAGE{{/C}}", chalk.red("MESSAGE"));
      verifyFormat("{{#C \"red\" \"yellow\"}}MESSAGE{{/C}}", chalk.red.yellow("MESSAGE"));
      verifyFormat("{{#C \"red yellow\"}}MESSAGE{{/C}}", chalk.red.yellow("MESSAGE"));
      verifyFormat("{{#C \"red\" yellow}}MESSAGE{{/C}}", chalk.red("MESSAGE"));
      verifyFormat("{{#C \"red\"}}MESSAGE{{#C \"yellow\"}}INNER{{/C}}RESTORE{{/C}}", chalk.red(`MESSAGE${chalk.yellow("INNER")}RESTORE`));
    });

    it("should concatenate styles, using characters differen than letters, numbers, dashes, pounds and underscores as separator", () => {
      verifyFormat("{{#C bgGreen.underline}}MESSAGE{{/C}}", chalk.bgGreen.underline("MESSAGE"));
      verifyFormat("{{#C bgGreen$underline}}MESSAGE{{/C}}", chalk.bgGreen.underline("MESSAGE"));
      verifyFormat("{{#C bgGreen@underline}}MESSAGE{{/C}}", chalk.bgGreen.underline("MESSAGE"));
      verifyFormat("{{#C bgGreen/underline}}MESSAGE{{/C}}", chalk.bgGreen.underline("MESSAGE"));
      verifyFormat("{{#C bgGreen.underline@bold}}MESSAGE{{/C}}", chalk.bgGreen.underline.bold("MESSAGE"));
    });

    it("should concatenate arguments, using the last as context", () => {
      verifyFormat(
        "{{#C red}}{{field}}{{/C}}", null, "{{#C green}}{{otherField}}{{/C}}", {field: "FIELD", otherField: "OTHER"},
        `${chalk.red("FIELD")}null${chalk.green("OTHER")}`
      );
    });
  });

  describe("error handling", () => {
    it("should return the source string unparsed when silent mode is enabled", () => {
      chalkbars.configuration.silent = true;
      verifyFormat("{{#C custom}}MESSAGE", "{{#C custom}}MESSAGE");
    });

    it("should throw the exception when silent mode is disabled", () => {
      chalkbars.configuration.silent = false;

      const wrapper = function(){
        chalkbars.format("{{#C custom}}MESSAGE");
      };

      expect(wrapper).to.throw(/Parse error on line 1/);

      chalkbars.configuration.silent = true;
    });
  });

  describe("color handling", () => {
    it("should handle 8-bit color codes", () => {
      verifyFormat("{{#C i#196 I#046}}MESSAGE{{/C}}", "\x1b[48;5;046m\x1b[38;5;196mMESSAGE\x1b[39m\x1b[49m");
      verifyFormat("{{#C x#500 X#050}}MESSAGE{{/C}}", "\x1b[48;5;46m\x1b[38;5;196mMESSAGE\x1b[39m\x1b[49m");
    });

    it("should handle 24-bit color codes", () => {
      verifyFormat("{{#C h#FF0000 H#00FF00}}MESSAGE{{/C}}", "\x1b[48;2;0;255;0m\x1b[38;2;255;0;0mMESSAGE\x1b[39m\x1b[49m");
    });
  });
});

describe("chalkbars.style", () => {
  it("should allow to fetch existing style", () => {
    expect(chalkbars.style()).to.not.exist;
    expect(chalkbars.style("")).to.not.exist;
    expect(chalkbars.style("bracket")).to.equal("bold blue");
  });

  it("should allow to fetch existing style", () => {
    chalkbars.style("new-style-1", "red");
    expect(chalkbars.style("new-style-1")).to.equal("red");
    chalkbars.style("new-style-1", "green");
    expect(chalkbars.style("new-style-1")).to.equal("green");
  });

  it("should allow to remove a style", () => {
    chalkbars.style("new-style-1", "red");
    chalkbars.style("new-style-2", "green");
    chalkbars.style("new-style-3", "blue");

    expect(chalkbars.style("new-style-1", null)).to.equal("red");
    expect(chalkbars.style("new-style-2", "")).to.equal("green");
    expect(chalkbars.style("new-style-3", false)).to.equal("blue");

    expect(chalkbars.style("new-style-1")).to.not.exist;
    expect(chalkbars.style("new-style-2")).to.not.exist;
    expect(chalkbars.style("new-style-3")).to.not.exist;
  });

  it("should NOT allow to shadow a chalk style", () => {
    const wrapper = function(){
      chalkbars.style("red", "yellow");
    };

    expect(wrapper).to.throw(/Cannot use "red" as a custom style name as it would shadow a chalk style/);
  });

  it("should NOT allow to define nested styles", () => {
    const wrapper = function(){
      chalkbars.style("new-style-2", "bracket");
    };

    expect(wrapper).to.throw(/Nesting of custom styles is not supported/);
  });

  it("should NOT allow values different from string", () => {
    const wrapper = function(){
      chalkbars.style("new-style-2", {a: 1});
    };

    expect(wrapper).to.throw(/Only strings are supported a values for styles/);
  });

  it("should resolve styles in templates", () => {
    verifyFormat("{{#C \"bracket-2\"}}MESSAGE{{/C}}", "MESSAGE");
    verifyFormat("{{#C \"bracket\"}}MESSAGE{{/C}}", chalk.bold.blue("MESSAGE"));
  });

  it("should have a list of default styles", () => {
    verifyFormat("{{#C \"highlight\"}}MESSAGE{{/C}}", "\x1b[1m\x1b[38;5;45mMESSAGE\x1b[39m\x1b[22m");
    verifyFormat("{{#C \"bracket\"}}MESSAGE{{/C}}", chalk.bold.blue("MESSAGE"));
    verifyFormat("{{#C \"info\"}}MESSAGE{{/C}}", chalk.bold.white("MESSAGE"));
    verifyFormat("{{#C \"skip\"}}MESSAGE{{/C}}", chalk.bold.gray("MESSAGE"));
    verifyFormat("{{#C \"ok\"}}MESSAGE{{/C}}", chalk.bold.green("MESSAGE"));
    verifyFormat("{{#C \"success\"}}MESSAGE{{/C}}", chalk.bold.green("MESSAGE"));
    verifyFormat("{{#C \"warn\"}}MESSAGE{{/C}}", chalk.bold.yellow("MESSAGE"));
    verifyFormat("{{#C \"fatal\"}}MESSAGE{{/C}}", chalk.bold.red("MESSAGE"));
    verifyFormat("{{#C \"fail\"}}MESSAGE{{/C}}", chalk.bold.red("MESSAGE"));
    verifyFormat("{{#C \"error\"}}MESSAGE{{/C}}", chalk.bold.red("MESSAGE"));
    verifyFormat("{{#C \"pass\"}}MESSAGE{{/C}}", chalk.bold.magenta("MESSAGE"));
    verifyFormat("{{#C \"debug\"}}MESSAGE{{/C}}", "\x1b[1m\x1b[38;5;127mMESSAGE\x1b[39m\x1b[22m");
  });
});

describe("chalkbars.formatNoColor", () => {
  it("should apply format and remove colors", () => {
    expect(chalkbars.formatNoColor("{{#C red}}{{field}}{{/C}}", {field: "FIELD"})).to.equal("FIELD");
    expect(chalkbars.formatNoColor("{{B warn}} Message {{E skip}}")).to.equal("[ WARN] Message \n\x1b[1A\x1b[74C[SKIP]");
  });
});

describe("chalkbars.plainFormat", () => {
  it("should apply format and remove all ANSI escapes", () => {
    expect(chalkbars.plainFormat("{{#C red}}{{field}}{{/C}}", {field: "FIELD"})).to.equal("FIELD");
    expect(chalkbars.plainFormat("{{B warn}} Message {{E skip}}")).to.equal("[ WARN] Message \n[SKIP]");
  });
});

describe("chalkbars.log", () => {
  it("should apply chalkbars and then log it to the console", () => {
    const stub = sinon.stub(console, "log");

    chalkbars.log("{{#C red}}{{field}}{{/C}}", {field: "FIELD"});
    chalkbars.log("{{B warn}} Message {{E skip}}");

    expect(stub.withArgs(chalk.red("FIELD")).calledOnce).to.be.true;
    expect(stub.withArgs(chalkbars.format("{{B warn}} Message {{E skip}}")).calledOnce).to.be.true;

    stub.restore();
  });
});

describe("chalkbars.error", () => {
  it("should apply chalkbars and then log it to the console error", () => {
    const stub = sinon.stub(console, "error");

    chalkbars.error("{{#C red}}{{field}}{{/C}}", {field: "FIELD"});
    chalkbars.error("{{B warn}} Message {{E skip}}");

    expect(stub.withArgs(chalk.red("FIELD")).calledOnce).to.be.true;
    expect(stub.withArgs(chalkbars.format("{{B warn}} Message {{E skip}}")).calledOnce).to.be.true;

    stub.restore();
  });
});

describe("chalkbars handlebars helpers", () => {
  describe("B", () => {
    it("should display a banner", () => {
      verifyFormat("{{B}}", "");
      verifyFormat("{{B other}}", chalk.bold.blue("[") + chalk.bold("OTHER") + chalk.bold.blue("]"));
      verifyFormat("{{B info}}", chalk.bold.blue("[") + chalk.bold.white(" INFO") + chalk.bold.blue("]"));
      verifyFormat("{{B info red}}", chalk.bold.blue("[") + chalk.bold.white.red("INFO RED") + chalk.bold.blue("]"));
      verifyFormat("{{B \"info\" \"red\"}}", chalk.bold.blue("[") + chalk.red(" INFO") + chalk.bold.blue("]"));
    });

    describe("should support common shortcuts", () => {
      const shortcuts = {BI: "info", BW: "warn", BO: "ok", BF: "fail", BE: "error", BP: "pass", BS: "skip", BD: "debug"};

      Object.keys(shortcuts).forEach(shortcut => {
        it(shortcut, () => {
          verifyFormat(`{{${shortcut}}}`, chalkbars.format(`{{B ${shortcuts[shortcut]}}}`));
        });
      });
    });
  });

  describe("E", () => {
    it("should display a footer", () => {
      verifyFormat("{{E}}", "");
      verifyFormat("{{E ok}}", `\n\x1b[1A\x1b[74C${chalk.bold.blue("[")}${chalk.bold.green(" OK ")}${chalk.bold.blue("]")}`);
      verifyFormat("{{E o}}", `\n\x1b[1A\x1b[74C${chalk.bold.blue("[")}${chalk.bold.green(" O  ")}${chalk.bold.blue("]")}`);
      verifyFormat("{{E ooo}}", `\n\x1b[1A\x1b[74C${chalk.bold.blue("[")}${chalk.bold.green(" OOO")}${chalk.bold.blue("]")}`);
      verifyFormat("{{E ok red}}", `\n\x1b[1A\x1b[72C${chalk.bold.blue("[")}${chalk.bold.green.red("OK RED")}${chalk.bold.blue("]")}`);
      verifyFormat("{{E \"ok\" \"red\"}}", `\n\x1b[1A\x1b[74C${chalk.bold.blue("[")}${chalk.red(" OK ")}${chalk.bold.blue("]")}`);
      verifyFormat(
        "This is another message {{E ok}}",
        `This is another message \n\x1b[1A\x1b[74C${chalk.bold.blue("[")}${chalk.bold.green(" OK ")}${chalk.bold.blue("]")}`
      );
      verifyFormat(
        "This is another message {{E \"warn\" \"green\"}}",
        `This is another message \n\x1b[1A\x1b[74C${chalk.bold.blue("[")}${chalk.green("WARN")}${chalk.bold.blue("]")}`
      );
    });

    describe("should support common shortcuts", () => {
      const shortcuts = {EI: "info", EW: "warn", EO: "ok", EF: "fail", EE: "error", EP: "pass", ES: "skip", ED: "debug"};

      Object.keys(shortcuts).forEach(shortcut => {
        it(shortcut, () => {
          verifyFormat(`{{${shortcut}}}`, chalkbars.format(`{{E ${shortcuts[shortcut]}}}`));
        });
      });
    });
  });

  it("B and E should use set brackets", () => {
    const oldOpening = chalkbars.configuration.openingBracket;
    const oldClosing = chalkbars.configuration.closingBracket;

    chalkbars.configuration.openingBracket = "(";
    chalkbars.configuration.closingBracket = ")";

    verifyFormat(
      "{{B warn}} Message {{E skip}}",
      `(${chalk.bold.yellow(" WARN")}) Message \n\x1b[1A\x1b[74C(${chalk.bold.gray("SKIP")})`
    );

    chalkbars.configuration.openingBracket = oldOpening;
    chalkbars.configuration.closingBracket = oldClosing;
  });
});

/* eslint-enable no-unused-expressions */


main.js
/*
 * This file is part of the chalkbars npm package. Copyright (C) 2015 and above Shogun <shogun@cowtech.it>.
 * Licensed under the MIT license, which can be found at http://www.opensource.org/licenses/mit-license.php.
 */

const chalk = require("chalk");
const templating = require("./lib/templating");
const configuration = require("./lib/configuration");
const style = require("./lib/functions").manageStyle;

/**
 * Compiles a Handlebars templates and then applies chalk colors.
 *
 * @alias module:chalkbars.format
 * @param {...string} template - The template to compile. You can specify more than one string, they will be concatenated.
 * @param {object} context - The context for the Handlebars template.
 * @returns {string} The compiled template with color styles applied.
 */
const format = function(){
  "use strict";

  const args = Array.prototype.slice.call(arguments, 0);

  if(!args.length)
    return "";

  // Get the last element and check whether is a object.
  let context = args.pop();

  if(!context || typeof context !== "object"){
    args.push(context);
    context = {};
  }

  // Perform the handlebars compilation
  try{
    return templating.renderTemplate(args, context);
  }catch(e){
    // If it fails we return the raw source or raise an error
    if(module.exports.configuration.silent)
      return args.join("");

    throw e;
  }
};

/**
 * Chalkbars module.
 *
 * @module chalkbars
 */
module.exports = {
  /**
   * The console library used by chalkbars.
   *
   * @type {object}
   */
  chalk: templating.chalk,

  /**
   * The templating library used by chalkbars.
   *
   * @type {object}
   */
  handlebars: templating.handlebars,

  configuration,

  style,

  format,

  /**
   * Compiles a Handlebars templates and then it strips out all ANSI color sequences.
   * {@see format}
   *
   * @param {...string} template - The template to compile. You can specify more than one string, they will be concatenated.
   * @param {object} context - The context for the Handlebars template.
   * @returns {string} The compiled template with color styles removed.
   */
  formatNoColor(){
    const args = Array.prototype.slice.call(arguments, 0);
    return format.apply(this, args).replace(/\u001b\[(?:[0-9]{1,3}(?:;[0-9]{1,3})*)?[m|K]/g, "");
  },

  /**
   * Compiles a Handlebars templates and then it strips out all ANSI escape sequences.
   * {@see format}
   *
   * @param {...string} template - The template to compile. You can specify more than one string, they will be concatenated.
   * @param {object} context - The context for the Handlebars template.
   * @returns {string} The compiled template with ANSI escape sequences removed.
   */
  plainFormat(){
    const args = Array.prototype.slice.call(arguments, 0);
    return chalk.stripColor(format.apply(this, args));
  },

  /**
   * Compiles a Handlebars templates and then outputs it to the console.
   * {@see format}
   *
   * @param {...string} template - The template to compile. You can specify more than one string, they will be concatenated.
   * @param {object} context - The context for the Handlebars template.
   */
  log(){
    const args = Array.prototype.slice.call(arguments, 0);
    console.log(format.apply(this, args));
  },

  /**
   * Compiles a Handlebars templates and then outputs it to the console as error.
   * {@see format}
   *
   * @param {...string} template - The template to compile. You can specify more than one string, they will be concatenated.
   * @param {object} context - The context for the Handlebars template.
   */
  error(){
    const args = Array.prototype.slice.call(arguments, 0);
    console.error(format.apply(this, args));
  }
};


lib/functions.js
/*
 * This file is part of the chalkbars npm package. Copyright (C) 2015 and above Shogun <shogun@cowtech.it>.
 * Licensed under the MIT license, which can be found at http://www.opensource.org/licenses/mit-license.php.
 */

const chalk = require("chalk");
const styles = require("./configuration").styles;

const HEX = 16;
const TO_RGB_R = 36;
const TO_RGB_G = 6;
const TO_RGB_TYPE = 5;
const FROM_RGB_R = 2;
const FROM_RGB_G = 4;
const FROM_RGB_B = 6;
const FOREGROUND_TAG = 38;
const BACKGROUND_TAG = 48;

module.exports = {
  parseStyles(raw){
    raw = raw.toString().replace(/[^a-z0-9#_-]/ig, " "); // Replace other characters with a space

    // Replace with custom styles
    return raw.replace(/\S+/ig, style => module.exports.manageStyle(style) || style);
  },

  embedInColor(subject, color, type, background){
    const tags = background ? [BACKGROUND_TAG, BACKGROUND_TAG + 1] : [FOREGROUND_TAG, FOREGROUND_TAG + 1];
    return `\x1b[${tags[0]};${type};${color}m${subject}\x1b[${tags[1]}m`;
  },

  embedIn256Color(subject, r, g, b, background){
    r = parseInt(r, 10) || 0;
    g = parseInt(g, 10) || 0;
    b = parseInt(b, 10) || 0;

    return module.exports.embedInColor(subject, HEX + (r * TO_RGB_R) + (g * TO_RGB_G) + b, TO_RGB_TYPE, background); // eslint-disable-line no-extra-parens
  },

  embedInHexColor(subject, color, background){
    const r = parseInt(color.substring(0, FROM_RGB_R), HEX) || 0;
    const g = parseInt(color.substring(FROM_RGB_R, FROM_RGB_G), HEX) || 0;
    const b = parseInt(color.substring(FROM_RGB_G, FROM_RGB_B), HEX) || 0;

    return module.exports.embedInColor(subject, `${r};${g};${b}`, 2, background);
  },

  /**
   * Get or sets a chalkbars style.
   *
   * @alias module:chalkbars.style
   * @param {string} name - The name of the style. It **must not** overwrite a existing chalk style.
   * @param {string|null|undefined} [value] -
   * * When omitted, it returns the current value of the style.
   * * When `null`, it returns the current value of the style and then deletes the style.
   * * When a string, it sets the new value of the style and then returns the value itself.
   *
   * @returns {string|undefined} The value of the style or `undefined`.
   */
  manageStyle(name, value){
    if(value === "")
      value = null;

    switch(typeof value){
      case "undefined": // Search the style
        value = styles[name];
        break;
      case "string": // Define a new style
        if(typeof chalk[name] === "function")
          throw new RangeError(`Cannot use "${name}" as a custom style name as it would shadow a chalk style.`);

        // Verify that the value does not contain other styles
        value = value.replace(/[^a-z0-9#_-]/ig, " ");

        value.split(/\s+/).forEach(style => {
          if(module.exports.manageStyle(style))
            throw new RangeError("Nesting of custom styles is not supported.");
        });

        styles[name] = value;
        break;
      default:
        // Null is handled here because typeof null == "object" -
        // If this if evaluates to true, then some truthy value different than a string was passed and this is not allow.
        // If this if evaluates to false, then all falsey values will result in deleting the style.
        if(value)
          throw new TypeError("Only strings are supported a values for styles.");

        // Delete a style. Return the current style as return value
        value = styles[name];
        delete styles[name]; // eslint-disable-line prefer-reflect
        break;
    }

    return value;
  }
};


lib/configuration.js
/*
 * This file is part of the chalkbars npm package. Copyright (C) 2015 and above Shogun <shogun@cowtech.it>.
 * Licensed under the MIT license, which can be found at http://www.opensource.org/licenses/mit-license.php.
 */

/* eslint-disable camelcase */

/**
 * Chalkbars configuration.
 *
 * @module chalkbars.configuration
 */
module.exports = {
  /**
   * The list of style recognized by chalkbars.
   * It can contain any chalk style, or the following special syntaxes:
   *
   * * _i#_**ABC** (where ABC is a 3 digit number): The ANSI color code. Example: **031** is red foreground, **042** is green background.
   * * _x#_**ABC** (where A, B, C are a number from 0 to 5): The 256 ANSI color code.
   *   For the meaning of the values, see {@link https://github.com/jbnicolai/ansi-256-colors}
   * * _X#_**AABBCC** (where AA, BB, CC ranges from 00 to FF): The RGB HEX color code.
   *
   * @type {object}
   */
  styles: {
    highlight: "bold x#045",
    // Normal styles
    bracket: "bold blue",
    info: "bold white",
    ok: "bold green",
    success: "bold green",
    warn: "bold yellow",
    fatal: "bold red",
    fail: "bold red",
    error: "bold red",
    pass: "bold magenta",
    debug: "bold x#303",
    skip: "bold gray",
    // The following style resembles the normal one but they don't use bold
    light_highlight: "x#045",
    light_bracket: "x#013",
    light_info: "white",
    light_ok: "green",
    light_success: "green",
    light_warn: "yellow",
    light_fatal: "red",
    light_fail: "red",
    light_error: "red",
    light_pass: "magenta",
    light_debug: "x#303",
    light_skip: "gray"
  },
  /**
   * Whether to ignore exceptions when compiling a Handlebars template fails. If set to `false` a compilation failure will raise an exception.
   *
   *
   *
   * @type boolean
   * @default true
   */
  silent: true,
  /**
   * The default template for the opening bracket in the `B` and `E` templates.
   *
   * @type string
   * @default {{#C bracket}}[{{/C}}
   */
  openingBracket: "{{#C bracket}}[{{/C}}",
  /**
   * The default template for the closing bracket in the `B` and `E` templates.
   *
   * @type string
   * @default {{#C bracket}}]{{/C}}
   */
  closingBracket: "{{#C bracket}}]{{/C}}",
  /**
   * The list of shortcuts available in the `B` and `E` templates.
   *
   * @type {object}
   */
  shortcuts: {
    I: "info",
    W: "warn",
    O: "ok",
    F: "fail",
    E: "error",
    P: "pass",
    S: "skip",
    D: "debug"
  }
};

/* eslint-enable camelcase */
